File: src/dotnet/UI.Blazor/Services/TuneUI/tune-ui.ts
import { delayAsync, PromiseSource } from 'actuallab-core';
import { getLogs } from 'logging';
import { DeviceInfo } from 'device-info';
import { SoundPlayer } from './sound-player';
import { Interactive } from 'interactive';
import { BrowserInfo } from '../BrowserInfo/browser-info';

const { logScope, debugLog, warnLog, errorLog } = getLogs('TuneUI');

// !!! keep in sync with TuneUI.cs
export enum Tune {
    None = 0,
    CancelReply,
    OpenModal,
    CloseModal,
    SelectNavbarItem,
    ShowInputError,
    BeginRecording,
    ConfirmRecording,
    EndRecording,
    RemindOfRecording,
    StartListening,
    StopListening,
    StartReplay,
    StopReplay,
    PinUnpinChat,
    NotifyOnNewMessageInApp,
    NotifyOnNewAudioMessageAfterDelay,
    SendMessage,
    EditMessage,
    ReplyMessage,
    ChangeAttachments,
    ChangeLanguage,
    ShowMenu,
    React,
    DragStart,
    ChangeToggle,
    ClickButton,
    WalkieReplyEnded,
    WalkieReplyNothingHeard,
    WalkieGestureDetected,
}

export type TuneName = keyof typeof Tune;

interface TuneInfo { vibration: number[], sound?: string }

const cooldownMap = new Map<Tune, number>([
    [Tune.CancelReply, 1],
    [Tune.OpenModal, 1],
    [Tune.CloseModal, 1],
    [Tune.SelectNavbarItem, 1],
    [Tune.ShowInputError, 1],
    [Tune.BeginRecording, 1],
    [Tune.ConfirmRecording, 1],
    [Tune.EndRecording, 1],
    [Tune.RemindOfRecording, 1],
    [Tune.StartListening, 1],
    [Tune.StopListening, 1],
    [Tune.StartReplay, 1],
    [Tune.StopReplay, 1],
    [Tune.NotifyOnNewMessageInApp, 5],
    [Tune.NotifyOnNewAudioMessageAfterDelay, 5],
    [Tune.SendMessage, 1],
    [Tune.EditMessage, 1],
    [Tune.ReplyMessage, 1],
    [Tune.ChangeAttachments, 1],
    [Tune.ChangeLanguage, 1],
    [Tune.ShowMenu, 1],
    [Tune.React, 1],
    [Tune.DragStart, 1],
]);

export class TuneUI {
    private static whenReady = new PromiseSource();
    private static blazorRef: DotNet.DotNetObject;
    private static tunes: Record<Tune, TuneInfo>;

    /** Called by blazor */
    public static init(blazorRef: DotNet.DotNetObject, tunes: Record<Tune, TuneInfo>) {
        this.blazorRef = blazorRef;
        this.tunes = tunes;
        this.whenReady.resolve(null);
    }

    /** Called by blazor */
    public static play(tune: Tune): void {
        void this.playAndWait(tune);
    }

    /** Called by blazor */
    public static async playAndWait(tune: Tune): Promise<void> {
        try {
            await this.whenReady;
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            const tuneInfo = this.tunes[tune] ?? this.tunes[Tune[tune]];
            // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
            if (!tuneInfo) {
                errorLog?.log(`${logScope}.playAndWait: unexpected tune ${tune}.`)
                return;
            }

            await Promise.all([this.playVibration(tune, tuneInfo), this.playSound(tune, tuneInfo)]);
        } catch (e) {
            warnLog?.log('Failed yo play tune', tune, e);
        }
    }

    // Private methods

    private static async playVibration(tune: Tune, tuneInfo: TuneInfo): Promise<void> {
        if (tuneInfo.vibration.length == 0)
            return;

        debugLog?.log(`playVibration: '${tune}'`);
        for (let i = 0; i < tuneInfo.vibration.length; i++) {
            const durationMs = tuneInfo.vibration[i];
            if (i % 2 == 0)
                await this.vibrateAndWait(durationMs);
            else
                await delayAsync(durationMs);
        }
    }

    private static async playSound(tune: Tune, tuneInfo: TuneInfo): Promise<void> {
        if (!tuneInfo.sound)
            return;

        debugLog?.log(`playSound: '${tune}'`);
        const ext = DeviceInfo.isWebKit ? '.m4a' : '.webm'; // TODO: allow webm for iOS >= 16.5
        const soundUrl = `dist/sounds/${tuneInfo.sound}${ext}`;
        const cooldown = cooldownMap.get(tune);
        if (BrowserInfo.hostKind !== 'MauiApp')
            await SoundPlayer.instance.play(soundUrl, cooldown);
        else
            await this.blazorRef.invokeMethodAsync('play', tune);
    }

    private static vibrate(durationMs = 20): void {
        const canVibrate = ('vibrate' in navigator);
        if (!canVibrate) {
            debugLog?.log(`vibrate(${durationMs}ms): unsupported by browser`);
            return;
        }

        if (!Interactive.isInteractive) {
            debugLog?.log(`vibrate(${durationMs}ms): suppressed (non-interactive yet)`);
            return;
        }

        if (!window.navigator.vibrate(durationMs)) {
            warnLog?.log(`vibrate(${durationMs}ms): suppressed by browser`);
            return;
        }

        debugLog?.log(`vibrate(${durationMs}ms)`);
    }

    private static vibrateAndWait(durationMs: number): Promise<void> {
        this.vibrate(durationMs);
        return delayAsync(durationMs);
    }
}

File: src/dotnet/UI.Blazor/Services/TuneUI/TuneUI.cs
using ActualChat.UI.Blazor.Module;
using ActualLab.Diagnostics;

namespace ActualChat.UI.Blazor.Services;

/// <summary>
/// Plays haptic feedback and sound effects for UI interactions.
/// </summary>
public abstract class TuneUI : ProcessorBase
{
    protected static readonly Dictionary<Tune, TuneInfo> Tunes = new () {
        // General actions
        [Tune.CancelReply] = new ([20] /*, "cancel-reply"*/),
        [Tune.OpenModal] = new ([20] /*, "open-modal"*/),
        [Tune.CloseModal] = new ([20] /*, "close-modal"*/),
        [Tune.SelectNavbarItem] = new ([20] /*, "select-navbar-item"*/),
        [Tune.ShowInputError] = new ([80] /*, "show-input-error"*/),
        [Tune.DragStart] = new ([100] /*, "drag-start"*/),
        [Tune.ChangeToggle] = new ([20] /*, "change-toggle"*/),
        [Tune.ClickButton] = new ([20] /*, "click-button"*/),
        // Recording
        [Tune.BeginRecording] = new ([100, 50, 50], "begin-recording"),
        [Tune.ConfirmRecording] = new ([50, 50, 100] /*, "confirm-recording"*/),
        [Tune.EndRecording] = new ([100], "end-recording"),
        [Tune.RemindOfRecording] = new ([20], "remind-of-recording"),
        // Playback
        [Tune.StartListening] = new ([100] /*, "start-listening"*/),
        [Tune.StopListening] = new ([20] /*, "stop-listening"*/),
        [Tune.StartReplay] = new ([100] /*, "start-replay"*/),
        [Tune.StopReplay] = new ([20] /*, "stop-replay"*/),
        // Chat UI
        [Tune.PinUnpinChat] = new ([50] /*, "pin-unpin-chat"*/),
        [Tune.NotifyOnNewMessageInApp] = new ([20], "notify-on-new-message-in-app"),
        [Tune.NotifyOnNewAudioMessageAfterDelay] = new ([20, 40, 100], "new-audio-message-after-delay"),
        [Tune.React] = new ([20, 10, 20]),
        // ChatMessageEditor
        [Tune.SendMessage] = new ([50] /*, "send-message"*/),
        [Tune.EditMessage] = new ([20] /*, "edit-message"*/),
        [Tune.ReplyMessage] = new ([20] /*, "reply-message"*/),
        [Tune.ChangeAttachments] = new ([20] /*, "change-attachments"*/),
        [Tune.ChangeLanguage] = new ([20, 20] /*, "change-language"*/),
        [Tune.ShowMenu] = new ([20] /*, "show-menu"*/),
        // Walkie-talkie
        [Tune.WalkieReplyEnded] = new ([100, 50, 100]),
        [Tune.WalkieReplyNothingHeard] = new ([80]),
    };
    // Suppress redundant tactile feedback fired as a side-effect of starting
    // recording: ConfirmRecording (mic-live signal ~0.5–1 s after start) and
    // StartListening (listening auto-enables for the recording chat) feel
    // like jitter on top of Tune.BeginRecording's initial haptic+sound.
    private static readonly TimeSpan BeginRecordingFollowupWindow = TimeSpan.FromMilliseconds(1500);
    private static readonly HashSet<Tune> BeginRecordingFollowups = [Tune.ConfirmRecording, Tune.StartListening];

    private static readonly string JSInitMethod = $"{BlazorUICoreModule.ImportName}.TuneUI.init";
    private DotNetObjectReference<TuneUI>? _backendRef;
    private CpuTimestamp _lastBeginRecordingAt;

    protected TuneUI(UIHub hub)
    {
        Hub = hub;
        _ = Initialize();
    }

    protected UIHub Hub { get; }

    protected ILogger Log => field ??= Hub.LogFor(GetType());
    protected ILogger? DebugLog => Log.IfEnabled(LogLevel.Information, Constants.DebugMode.Tunes);

    private async ValueTask Initialize()
    {
        try {
            _backendRef ??= DotNetObjectReference.Create(this);
            await Hub.JS.InvokeVoidAsync(JSInitMethod, _backendRef, Tunes).ConfigureAwait(false);
        }
        catch (Exception e) {
            Log.LogError(e, "Initialize failed");
        }
    }

    protected override Task DisposeAsyncCore()
    {
        _backendRef.DisposeSilently();
        return Task.CompletedTask;
    }

    public Task Play(Tune tune)
    {
        if (IsSuppressed(tune))
            return Task.CompletedTask;

        OnBeforePlay(tune);
        return PlayInternal(tune);
    }

    public Task PlayAndWait(Tune tune)
    {
        if (IsSuppressed(tune))
            return Task.CompletedTask;

        OnBeforePlay(tune);
        return PlayAndWaitInternal(tune);
    }

    protected abstract Task PlayInternal(Tune tune);

    protected abstract Task PlayAndWaitInternal(Tune tune);

    private bool IsSuppressed(Tune tune)
        => BeginRecordingFollowups.Contains(tune)
            && _lastBeginRecordingAt.Elapsed < BeginRecordingFollowupWindow;

    private void OnBeforePlay(Tune tune)
    {
        if (tune == Tune.BeginRecording)
            _lastBeginRecordingAt = CpuTimestamp.Now;
    }
}

/// <summary>
/// Identifies specific UI feedback sounds and haptic patterns.
/// </summary>
// !!! keep in sync with tune-ui.ts
public enum Tune
{
    None = 0,
    CancelReply,
    OpenModal,
    CloseModal,
    SelectNavbarItem,
    ShowInputError,
    BeginRecording,
    ConfirmRecording,
    EndRecording,
    RemindOfRecording,
    StartListening,
    StartReplay,
    StopReplay,
    StopListening,
    PinUnpinChat,
    NotifyOnNewMessageInApp,
    NotifyOnNewAudioMessageAfterDelay,
    SendMessage,
    EditMessage,
    ReplyMessage,
    ChangeAttachments,
    ChangeLanguage,
    ShowMenu,
    React,
    DragStart,
    ChangeToggle,
    ClickButton,
    WalkieReplyEnded,
    WalkieReplyNothingHeard,
}

/// <summary>
/// Defines vibration pattern and optional sound for a <see cref="Tune"/>.
/// </summary>
public record TuneInfo(int[] Vibration, string Sound = "");
